### PR TITLE
ISPN-11512 The mass indexer should iterate over the cache only once

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.redhat.com/browse/ISPN-11512

Performs the mass indexer grouping same indexes types (shared/non-shared)

### CODE REVIEW

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 1]
```diff
@@ -151,7 +153,15 @@ private Object extractValue(Object storageValue, DataConversion valueDataConvers
       @Override
       public void writeObject(ObjectOutput output, IndexWorker worker) throws IOException {
          output.writeObject(worker.cacheName);
-         output.writeObject(PojoIndexedTypeIdentifier.convertToLegacy(worker.indexedType));
+         if (worker.indexedTypes == null) {
+            output.writeInt(0);
+         } else {
+            Set<? extends Class<?>> types = worker.indexedTypes.stream().map(PojoIndexedTypeIdentifier::convertToLegacy).collect(Collectors.toSet());
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: This ```types``` is used only to count them ? And then writeObject + convertToLegacy again? surely a mistake

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 2]
```diff
@@ -162,14 +172,19 @@ public void writeObject(ObjectOutput output, IndexWorker worker) throws IOExcept
       @Override
       public IndexWorker readObject(ObjectInput input) throws IOException, ClassNotFoundException {
          String cacheName = (String) input.readObject();
-         Class indexedClass = (Class) input.readObject();
+         int typesSize = input.readInt();
+         Set<IndexedTypeIdentifier> types = new HashSet<>(typesSize);
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: Is the initial capacity really helping ?

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 3]
```diff
@@ -162,14 +172,19 @@ public void writeObject(ObjectOutput output, IndexWorker worker) throws IOExcept
...
+         if (typesSize > 0) {
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: The ```if (typesSize > 0) ``` is not needed, the ```for``` already guards against that.

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 4]
```diff
@@ -46,17 +47,17 @@
...
-   private final IndexedTypeIdentifier indexedType;
+   private final Set<IndexedTypeIdentifier> indexedTypes;
    private final boolean flush;
    private final boolean clean;
    private final boolean skipIndex;
    private final boolean primaryOwner;
    private final Set<Object> keys;
 
-   IndexWorker(String cacheName, IndexedTypeIdentifier indexedType, boolean flush, boolean clean, boolean skipIndex,
+   IndexWorker(String cacheName, Set<IndexedTypeIdentifier> indexedTypes, boolean flush, boolean clean, boolean skipIndex,
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: Perfect timing to also sort out the TODO on line 76

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 5]
```diff
@@ -151,7 +153,15 @@ private Object extractValue(Object storageValue, DataConversion valueDataConvers
       @Override
       public void writeObject(ObjectOutput output, IndexWorker worker) throws IOException {
          output.writeObject(worker.cacheName);
-         output.writeObject(PojoIndexedTypeIdentifier.convertToLegacy(worker.indexedType));
+         if (worker.indexedTypes == null) {
+            output.writeInt(0);
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: When reading, this silently leads to null being transformed into empty set. Do we care?

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java [hunk 6]
```diff
@@ -103,8 +104,9 @@ public Void apply(EmbeddedCacheManager embeddedCacheManager) {
                      value = wrapper.wrap(value);
                   }
                   //TODO do not use Class equality but refactor to type equality:
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: I think this TODO was just resolved :grin:

File: query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java [hunk 7]
```diff
@@ -186,33 +186,32 @@ public boolean isRunning() {
...
             try {
-               for (IndexedTypeIdentifier type : toFlush) {
-                  indexUpdater.flush(type);
-               }
+               indexUpdater.flush(toFlush);
             } finally {
                lock.unlock();
             }
          };
          CompletableFuture<Void> compositeFuture;
```
Comment (anistor) [inline query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java]: This ```compositeFuture``` could be declared directly where it is used.

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java (hunk 1 above)
Comment (gustavonalle) [reply query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: ops...

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java (hunk 4 above)
Comment (gustavonalle) [reply query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: I have upcoming PRs that will get rid of mostly of that decoration

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java (hunk 5 above)
Comment (gustavonalle) [reply query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: not really, when MassIndexing only a set of keys, the types are ignored

File: query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java (hunk 2 above)
Comment (gustavonalle) [reply query/src/main/java/org/infinispan/query/impl/massindex/IndexWorker.java]: TBH using or not initial capacity here is irrelevant: this Set will be generally small, and will have be created almost never

**Comment by anistor** [comment]
Merged in master. Thanks @gustavonalle !

**Pull request merged**